File: venus-sealer/sealing/mock.go
package sealing

import (
	"context"
	"math/rand"
	"sync"
	"sync/atomic"

	"github.com/filecoin-project/go-state-types/abi"

	"github.com/dtynn/venus-cluster/venus-sealer/logging"
	"github.com/dtynn/venus-cluster/venus-sealer/sealing/api"
)

var _ api.SealerAPI = (*Mock)(nil)

func NewMock(miner abi.ActorID, proofType abi.RegisteredSealProof) (*Mock, error) {
	m := &Mock{
		miner:        miner,
		sectorNumber: 0,
		proofType:    proofType,
		ticket: api.Ticket{
			Ticket: make([]byte, 32),
			Epoch:  1,
		},
		seed: api.Seed{
			Seed:  make([]byte, 32),
			Epoch: 1,
		},

		l: log.With("mod", "mock"),
	}

	rand.Read(m.ticket.Ticket[:])
	rand.Read(m.seed.Seed[:])

	m.preCommits.commits = map[abi.SectorID]api.PreCommitOnChainInfo{}
	m.proofs.proofs = map[abi.SectorID]api.ProofOnChainInfo{}

	return m, nil
}

type Mock struct {
	miner        abi.ActorID
	sectorNumber uint64
	proofType    abi.RegisteredSealProof
	ticket       api.Ticket
	seed         api.Seed

	l *logging.ZapLogger

	preCommits struct {
		sync.RWMutex
		commits map[abi.SectorID]api.PreCommitOnChainInfo
	}

	proofs struct {
		sync.RWMutex
		proofs map[abi.SectorID]api.ProofOnChainInfo
	}
}

func (m *Mock) AllocateSector(ctx context.Context, spec api.AllocateSectorSpec) (*api.AllocatedSector, error) {
	m.l.Infof("allocate sector: %#v", spec)

	minerFits := len(spec.AllowedMiners) == 0
	for _, want := range spec.AllowedMiners {
		if want == m.miner {
			minerFits = true
			break
		}
	}

	if !minerFits {
		return nil, nil
	}

	typeFits := len(spec.AllowedProofTypes) == 0
	for _, typ := range spec.AllowedProofTypes {
		if typ == m.proofType {
			typeFits = true
			break
		}
	}

	if !typeFits {
		return nil, nil
	}

	next := atomic.AddUint64(&m.sectorNumber, 1)
	return &api.AllocatedSector{
		ID: abi.SectorID{
			Miner:  m.miner,
			Number: abi.SectorNumber(next),
		},
		ProofType: m.proofType,
	}, nil
}

func (m *Mock) AcquireDeals(ctx context.Context, sid abi.SectorID, spec api.AcquireDealsSpec) (api.Deals, error) {
	m.l.Infof("acquire deals for %d: %#v", sid, spec)
	return nil, nil
}

func (m *Mock) AssignTicket(ctx context.Context, sid abi.SectorID) (api.Ticket, error) {
	m.l.Infof("assign ticket for %d", sid)
	return m.ticket, nil
}

func (m *Mock) SubmitPreCommit(ctx context.Context, sector api.AllocatedSector, info api.PreCommitOnChainInfo) (api.SubmitPreCommitResp, error) {
	m.l.Infof("submit pre commit: %v", info.CommR)

	m.preCommits.Lock()
	defer m.preCommits.Unlock()

	if _, ok := m.preCommits.commits[sector.ID]; ok {
		return api.SubmitPreCommitResp{
			Res:  api.SubmitDuplicateSubmit,
			Desc: nil,
		}, nil
	}

	m.preCommits.commits[sector.ID] = info

	return api.SubmitPreCommitResp{
		Res:  api.SubmitAccepted,
		Desc: nil,
	}, nil
}

func (m *Mock) PollPreCommitState(ctx context.Context, sid abi.SectorID) (api.PollPreCommitStateResp, error) {
	m.l.Infof("poll pre commit state for %d", sid)

	m.preCommits.RLock()
	defer m.preCommits.RUnlock()

	if _, ok := m.preCommits.commits[sid]; ok {
		return api.PollPreCommitStateResp{
			State: api.OnChainStateLanded,
			Desc:  nil,
		}, nil
	}

	return api.PollPreCommitStateResp{
		State: api.OnChainStateNotFound,
		Desc:  nil,
	}, nil
}

func (m *Mock) AssignSeed(ctx context.Context, sid abi.SectorID) (api.Seed, error) {
	m.l.Infof("assign seed for %d", sid)
	return m.seed, nil
}

func (m *Mock) SubmitProof(ctx context.Context, id abi.SectorID, info api.ProofOnChainInfo) (api.SubmitProofResp, error) {
	m.l.Infof("submit proof: %v", info.Proof)

	m.proofs.Lock()
	defer m.proofs.Unlock()

	if _, ok := m.proofs.proofs[id]; ok {
		return api.SubmitProofResp{
			Res:  api.SubmitDuplicateSubmit,
			Desc: nil,
		}, nil
	}

	m.proofs.proofs[id] = info

	return api.SubmitProofResp{
		Res:  api.SubmitAccepted,
		Desc: nil,
	}, nil
}

func (m *Mock) PollProofState(ctx context.Context, sid abi.SectorID) (api.PollProofStateResp, error) {
	m.l.Infof("assign seed for %d", sid)

	m.proofs.RLock()
	defer m.proofs.RUnlock()

	if _, ok := m.proofs.proofs[sid]; ok {
		return api.PollProofStateResp{
			State: api.OnChainStateLanded,
			Desc:  nil,
		}, nil
	}

	return api.PollProofStateResp{
		State: api.OnChainStateNotFound,
		Desc:  nil,
	}, nil
}

File: venus-worker/src/rpc/sealer/mock.rs
//! provides mock impl for the SealerRpc

use std::collections::HashMap;
use std::sync::{
    atomic::{AtomicU64, Ordering},
    RwLock,
};

use fil_types::ActorID;
use jsonrpc_core::Error;

use super::*;
use crate::{
    logging::{debug_field, error, warn},
    types::SealProof,
};

/// simplest mock server implementation
pub struct SimpleMockSealerRpc {
    miner: ActorID,
    sector_number: AtomicU64,
    proof_type: SealProof,
    ticket: Ticket,
    seed: Seed,

    pre_commits: RwLock<HashMap<SectorID, PreCommitOnChainInfo>>,
    proofs: RwLock<HashMap<SectorID, ProofOnChainInfo>>,
}

impl SimpleMockSealerRpc {
    /// constructs a SimpleMockSealerRpc with given miner & seal proof type
    pub fn new(miner: ActorID, proof_type: SealProof) -> Self {
        SimpleMockSealerRpc {
            miner,
            sector_number: Default::default(),
            proof_type,
            ticket: Default::default(),
            seed: Default::default(),
            pre_commits: RwLock::new(Default::default()),
            proofs: RwLock::new(Default::default()),
        }
    }
}

impl Sealer for SimpleMockSealerRpc {
    fn allocate_sector(&self, spec: AllocateSectorSpec) -> Result<Option<AllocatedSector>> {
        if let Some(false) = spec
            .allowed_miners
            .as_ref()
            .map(|miners| miners.contains(&self.miner))
        {
            return Ok(None);
        }

        if let Some(false) = spec
            .allowed_proof_types
            .as_ref()
            .map(|types| types.contains(&self.proof_type))
        {
            return Ok(None);
        }

        let next = self.sector_number.fetch_add(1, Ordering::SeqCst);

        Ok(Some(AllocatedSector {
            id: SectorID {
                miner: self.miner,
                number: next,
            },
            proof_type: self.proof_type,
        }))
    }

    fn acquire_deals(&self, _id: SectorID, _spec: AcquireDealsSpec) -> Result<Option<Deals>> {
        Ok(None)
    }

    fn assign_ticket(&self, _id: SectorID) -> Result<Ticket> {
        Ok(self.ticket.clone())
    }

    fn submit_pre_commit(
        &self,
        sector: AllocatedSector,
        info: PreCommitOnChainInfo,
        reset: bool,
    ) -> Result<SubmitPreCommitResp> {
        let mut pre_commits = self.pre_commits.write().map_err(|e| {
            error!(err = debug_field(&e), "acquire write lock");
            Error::internal_error()
        })?;

        if !reset {
            if let Some(_exist) = pre_commits.get(&sector.id) {
                return Ok(SubmitPreCommitResp {
                    res: SubmitResult::DuplicateSubmit,
                    desc: None,
                });
            }
        }

        pre_commits.insert(sector.id, info);
        Ok(SubmitPreCommitResp {
            res: SubmitResult::Accepted,
            desc: None,
        })
    }

    fn poll_pre_commit_state(&self, id: SectorID) -> Result<PollPreCommitStateResp> {
        let pre_commits = self.pre_commits.read().map_err(|e| {
            error!(err = debug_field(&e), "acquire read lock");
            Error::internal_error()
        })?;

        match pre_commits.get(&id) {
            Some(_) => Ok(PollPreCommitStateResp {
                state: OnChainState::Landed,
                desc: None,
            }),

            None => Ok(PollPreCommitStateResp {
                state: OnChainState::NotFound,
                desc: None,
            }),
        }
    }

    fn submit_persisted(&self, id: SectorID, instance: String) -> Result<()> {
        warn!(
            "sector m-{}-s-{} has been persisted into instance {}",
            id.miner, id.number, instance
        );
        return Ok(());
    }

    fn wait_seed(&self, _id: SectorID) -> Result<WaitSeedResp> {
        Ok(WaitSeedResp {
            should_wait: false,
            delay: 0,
            seed: Some(self.seed.clone()),
        })
    }

    fn submit_proof(
        &self,
        id: SectorID,
        proof: ProofOnChainInfo,
        reset: bool,
    ) -> Result<SubmitProofResp> {
        let mut proofs = self.proofs.write().map_err(|e| {
            error!(err = debug_field(&e), "acquire write lock");
            Error::internal_error()
        })?;

        if !reset {
            if let Some(_exist) = proofs.get(&id) {
                return Ok(SubmitProofResp {
                    res: SubmitResult::DuplicateSubmit,
                    desc: None,
                });
            }
        }

        proofs.insert(id, proof);
        Ok(SubmitProofResp {
            res: SubmitResult::Accepted,
            desc: None,
        })
    }

    fn poll_proof_state(&self, id: SectorID) -> Result<PollProofStateResp> {
        let proofs = self.proofs.read().map_err(|e| {
            error!(err = debug_field(&e), "acquire read lock");
            Error::internal_error()
        })?;

        match proofs.get(&id) {
            Some(_) => Ok(PollProofStateResp {
                state: OnChainState::Landed,
                desc: None,
            }),

            None => Ok(PollProofStateResp {
                state: OnChainState::NotFound,
                desc: None,
            }),
        }
    }

    fn report_state(&self, id: SectorID, req: ReportStateReq) -> Result<()> {
        warn!(miner = id.miner, num = id.number, "report state: {:?}", req);
        Ok(())
    }

    fn report_finalized(&self, id: SectorID) -> Result<()> {
        warn!(miner = id.miner, num = id.number, "report finalized");
        Ok(())
    }
}
